### PR TITLE
[release-0.18] Reorders function declarations in JuMPArray.jl

### LINKED ISSUE #1531: Reorder function declarations in JuMPArray.jl to improve compatibility with PackageCompiler
As described in [this thread](https://discourse.julialang.org/t/change-to-improve-jump-compatibility-with-packagecompiler/15944) on Discourse, I've found the order of function declarations in JuMPArray.jl (release 0.18) creates an incompatibility with PackageCompiler. The problem is there are @generated functions (Base.getindex and Base.setindex!) that reference a regular function (_to_cartesian) before it is declared. This causes an error when using PackageCompiler.

Moving the declaration statements to the end of the JuMPArray.jl file fixes the problem.

### PR DESCRIPTION
**Pull request by jwveysey**
Reorders function declarations in JuMPArray.jl to improve compatibility with PackageCompiler. Adds tests of compiling JuMP into a new system image using PackageCompiler.

Closes #1531.

### CODE REVIEW

**Comment by jwveysey** [comment]
Hi, all - here's my attempt at a pull request to address issue #1531. Comments appreciated. Mostly what's new here are some tests of compiling JuMP with PackageCompiler. These take a while to execute due to the overhead of running PackageCompiler and creating a new system image. I'm not sure there's any way around that if you want to test the compilation _per se_.

As you may know, PackageCompiler requires a set of "snooping" commands that exercise the package that's being compiled. The JIT compilation that results from these commands is inspected to create a set of instructions for the ahead-of-time compilation. I chose snooping commands that create and solve LP and MILP models (borrowing from existing tests) since that's how I've been using PackageCompiler in reality. The commands could be extended to cover other use cases if desired.

**Comment by blegat** [comment]
You need to add PackageCompiler in `test/REQUIRE`

**Comment by jwveysey** [comment]
Quite right – sorry about that! I just made the necessary change.

**Comment by IainNZ** [comment]
How long does the new test take to run?

**Comment by blegat** [comment]
~ 30 min

**Comment by mlubin** [comment]
Seems to be failing on 0.6 and 1.0?

**Comment by ianfiske** [comment]
Had a look at the 1.0 failure.  The error is `UndefVarError: serialize not defined`.  @jwveysey You just need to put `using Serialization` in your test.

**Comment by jwveysey** [comment]
Thanks for tracking it down, ianfiske.  I believe I've also identified the problem in v. 0.6.  I'm going to try to fix both today.

**Comment by blegat** [comment]
Looking at the log we have: https://travis-ci.org/JuliaOpt/JuMP.jl/jobs/440289533#L3809-L3811
Why is it complaining about undefined variable "MathOptInterface" ? Why is it using MathOptInterface on JuMP release-0.18 branch ?

**Comment by jwveysey** [comment]
I'm not sure about that. The code I added doesn't call MathOptInterface directly, so I assume it's being referenced somewhere in JuMP itself.

**Comment by blegat** [comment]
It is used by some solvers. Aren't these warning complaining about undefined variables an issue ?

**Comment by jwveysey** [comment]
Having just taken a closer look at PackageCompiler's code, I would say not.  In this case, it's PackageCompiler working as designed.  I believe what's happening is the following:

1) PackageCompiler uses the snooping commands (the script in snoop_commands.jl) to run JuMP through some use cases and to create a precompilation file.  This file contains a set of precompile() statements based on the use cases.  Many of these statements reference JuMP, and some reference other packages involved in the use cases.  

2) At the beginning of the precompilation file, PackageCompiler inserts commands to import JuMP and any packages referenced in JuMP.  This behavior is not user-modifiable.

3) PackageCompiler builds a new system image using the precompilation file (among other resources).  The build operation happens in a separate instance of Julia, so the operative namespace contains only the packages in 2).

4) Every precompile() statement is wrapped in a try-catch block, however, so statements referencing packages not included in the current namespace are skipped.  These include statements that refer to other packaged involved in the use cases but not imported in JuMP itself.

The end result is probably a system image that's not quite as efficient for the test cases as it could be.  But it's still much better than the default system image.  And in any event, the exclusion of packages referenced in the use cases but not in the package being compiled is an inherent limitation of PackageCompiler.

**Comment by jwveysey** [comment]
Noted, thanks. I've just implemented the requested change. The compile test script (compilation.jl) has been removed from runtests.jl and can be run separately as desired.

**Comment by jwveysey** [comment]
@mlubin, please let me know if anything else is needed to accept this pull request. Thanks!

File: test/REQUIRE [hunk 1]
```diff
@@ -7,3 +7,4 @@ Ipopt
 ECOS
 SCS
 OffsetArrays 0.2.13
+PackageCompiler
```
Comment (odow) [inline test/REQUIRE]: Since this isn't getting run in the automated tests, we probably don't need this in REQUIRE?

**Pull request merged**